File: Tutorials/OpenGLLayerExample/netcore/UniformSender.cs
using System;
using System.Drawing;


// GlmNet can be found at https://github.com/dwmkerr/glmnet
using GlmNet;

using static OpenGLLayerExample.OGLProvider;

namespace OpenGLLayerExample
{
    public static class UniformSender
    {
        public static void SendColorAsUniformVariable(uint shader, Color fill, string name)
        {
            float[] argb = { (float)fill.R / 255.0f, (float)fill.G / 255.0f, (float)fill.B / 255.0f };
            int location = GL.GetUniformLocation(shader, name);
            if (location < 0)
                Console.WriteLine("Uniform <" + name + "> not found.");
            GL.Uniform3(location, argb[0], argb[1], argb[2]);
        }

        public static void SendVec3AsUniformVariable(uint shader, float[] value, string name)
        {
            int location = GL.GetUniformLocation(shader, name);
            if (location < 0)
                Console.WriteLine("Uniform <" + name + "> not found.");
            GL.Uniform3(location, value[0], value[1], value[2]);
        }

        public static void SendMVPAsUniformVariable(uint shader, mat4 model, mat4 view, mat4 projection)
        {
            int location = GL.GetUniformLocation(shader, "model");
            if (location < 0)
                Console.WriteLine("Uniform <model> not found.");
            GL.UniformMatrix4(location, 1, false, model.to_array());

            location = GL.GetUniformLocation(shader, "view");
            if (location < 0)
                Console.WriteLine("Uniform <view> not found.");
            GL.UniformMatrix4(location, 1, false, view.to_array());

            location = GL.GetUniformLocation(shader, "projection");
            if (location < 0)
                Console.WriteLine("Uniform <projection> not found.");
            GL.UniformMatrix4(location, 1, false, projection.to_array());
        }

        public static void SendUniformSample2D(uint shader, string name)
        {
            int location = GL.GetUniformLocation(shader, name);
            if (location < 0)
                Console.WriteLine("Uniform <" + name + "> not found.");

            GL.Uniform1(location, 0);
        }
    }
}

File: Tutorials/OpenGLLayerExample/java/src/main/java/OpenGLLayerExample/UniformSender.java
package OpenGLLayerExample;

import java.awt.Color;
import java.nio.FloatBuffer;

import org.lwjgl.BufferUtils;

import static org.lwjgl.opengl.GL20.*;

//GLM can be found at https://mvnrepository.com/artifact/io.github.java-graphics/glm/1.0.1
import glm.mat._4.Mat4;

public final class UniformSender {

    private UniformSender() {
    }

    public static void sendColorAsUniformVariable(int shader, Color fill, String name) {
        float[] argb = { (float) fill.getRed() / 255.0f, (float) fill.getGreen() / 255.0f,
                (float) fill.getBlue() / 255.0f };

        int location = glGetUniformLocation(shader, name);
        if (location < 0)
            System.out.println("Uniform <" + name + "> not found.");

        glUniform3f(location, argb[0], argb[1], argb[2]);
    }

    public static void sendVec3AsUniformVariable(int shader, float[] value, String name) {
        int location = glGetUniformLocation(shader, name);
        if (location < 0)
            System.out.println("Uniform <" + name + "> not found.");
        glUniform3f(location, value[0], value[1], value[2]);
    }

    public static void sendMVPAsUniformVariable(int shader, Mat4 model, Mat4 view, Mat4 projection) {
        int location = glGetUniformLocation(shader, "model");
        if (location < 0)
            System.out.println("Uniform <model> not found.");
        glUniformMatrix4fv(location, false, toFloatBuffer(model));

        location = glGetUniformLocation(shader, "view");
        if (location < 0)
            System.out.println("Uniform <view> not found.");
        glUniformMatrix4fv(location, false, toFloatBuffer(view));

        location = glGetUniformLocation(shader, "projection");
        if (location < 0)
            System.out.println("Uniform <projection> not found.");
        glUniformMatrix4fv(location, false, toFloatBuffer(projection));
    }

    public static void sendUniformSample2D(int shader, String name) {
        int location = glGetUniformLocation(shader, name);
        if (location < 0)
            System.out.println("Uniform <" + name + "> not found.");

        glUniform1i(location, 0);
    }

    private static FloatBuffer toFloatBuffer(Mat4 mat) {
        FloatBuffer bb = BufferUtils.createFloatBuffer(4 * 4);
        float[] array = mat.toFa_();
        bb.put(array);
        bb.rewind();
        return bb;
    }
}
